linalg/batched/gpu_blas.cpp

```diff
@@ -0,0 +1,195 @@
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+
+#include "gpu_blas.hpp"
+#include "../../general/forall.hpp"
+
+#ifdef MFEM_USE_CUDA_OR_HIP
+
+#if defined(MFEM_USE_CUDA)
+#define MFEM_cu_or_hip(stub) cu##stub
+#define MFEM_CU_or_HIP(stub) CU##stub
+#elif defined(MFEM_USE_HIP)
+#define MFEM_cu_or_hip(stub) hip##stub
+#define MFEM_CU_or_HIP(stub) HIP##stub
+#endif
+
+#define MFEM_CONCAT(x, y, z) MFEM_CONCAT_(x, y, z)
+#define MFEM_CONCAT_(x, y, z) x ## y ## z
+
+#ifdef MFEM_USE_SINGLE
+#define MFEM_GPUBLAS_PREFIX(stub) MFEM_CONCAT(MFEM_cu_or_hip(blas), S, stub)
+#elif defined(MFEM_USE_DOUBLE)
+#define MFEM_GPUBLAS_PREFIX(stub) MFEM_CONCAT(MFEM_cu_or_hip(blas), D, stub)
+#endif
+
+#define MFEM_BLAS_SUCCESS MFEM_CU_or_HIP(BLAS_STATUS_SUCCESS)
+
+#if defined(MFEM_USE_CUDA)
+#include <cublas_v2.h>
+#endif
+
+namespace mfem
+{
+
+using blasStatus_t = MFEM_cu_or_hip(blasStatus_t);
+
+GPUBlas::GPUBlas()
+{
+   blasStatus_t status = MFEM_cu_or_hip(blasCreate)(&handle);
+   MFEM_VERIFY(status == MFEM_BLAS_SUCCESS, "Cannot initialize GPU BLAS.");
+}
+
+GPUBlas::~GPUBlas()
+{
+   MFEM_cu_or_hip(blasDestroy)(handle);
+}
+
+GPUBlas &GPUBlas::Instance()
+{
+   static GPUBlas instance;
+   return instance;
+}
+
+MFEM_cu_or_hip(blasHandle_t) GPUBlas::Handle()
+{
+   return Instance().handle;
+}
+
+void GPUBlas::EnableAtomics()
+{
+   const blasStatus_t status = MFEM_cu_or_hip(blasSetAtomicsMode)(
+                                  Handle(), MFEM_CU_or_HIP(BLAS_ATOMICS_ALLOWED));
+   MFEM_VERIFY(status == MFEM_BLAS_SUCCESS, "GPU BLAS error.");
+}
+
+void GPUBlas::DisableAtomics()
+{
+   const blasStatus_t status = MFEM_cu_or_hip(blasSetAtomicsMode)(
+                                  Handle(), MFEM_CU_or_HIP(BLAS_ATOMICS_NOT_ALLOWED));
+   MFEM_VERIFY(status == MFEM_BLAS_SUCCESS, "GPU BLAS error.");
+}
+
+void GPUBlasBatchedLinAlg::AddMult(const DenseTensor &A, const Vector &x,
+                                   Vector &y, real_t alpha, real_t beta) const
+{
+   const int m = A.SizeI();
+   const int n = A.SizeJ();
+   const int n_mat = A.SizeK();
+   const int k = x.Size() / n / n_mat;
+
+   auto d_A = mfem::Reshape(A.Read(), m, n, n_mat);
+   auto d_x = mfem::Reshape(x.Read(), n, k, n_mat);
+   auto d_y = mfem::Reshape(beta == 0.0 ? y.Write() : y.ReadWrite(), m, k, n_mat);
+
+   const auto op = MFEM_CU_or_HIP(BLAS_OP_N);
+
+   const blasStatus_t status = MFEM_GPUBLAS_PREFIX(gemmStridedBatched)(
+                                  GPUBlas::Handle(), op, op, m, k, n, &alpha,
+                                  d_A, m, m*n, d_x, n, n*k, &beta, d_y, m, m*k,
+                                  n_mat);
+   MFEM_VERIFY(status == MFEM_BLAS_SUCCESS, "GPU BLAS error.");
+}
+
+void GPUBlasBatchedLinAlg::LUFactor(DenseTensor &A, Array<int> &P) const
+{
+   const int n = A.SizeI();
+   const int n_mat = A.SizeK();
+
+   P.SetSize(n*n_mat);
```

Comment:
If we can configure the member containers with memory type and a "release memory" call that would be helpful in applications.

Reply:
The pivot array `P` is an input here, so the user can configure the memory type. There are some internal arrays that contain the pointers. Once something like a "workspace allocator" (#4065) is available, we should use that.